lib/src/analyzer/parser.dart

```diff
@@ -0,0 +1,166 @@
+import 'package:analyzer/dart/ast/ast.dart' as ast;
+import 'package:analyzer/source/line_info.dart';
+import 'package:testgen/src/analyzer/declaration.dart';
+
+const int failedToExtractID = -1;
+
+List<Declaration> parseCompilationUnit(ast.CompilationUnit unit, String path) {
+  final declarations = <Declaration>[];
+  final lineInfo = unit.lineInfo;
+
+  for (final member in unit.declarations) {
+    switch (member) {
+      case ast.ClassDeclaration():
+        declarations.addAll(_parseClassDeclaration(member, lineInfo, path));
+        break;
+      case ast.FunctionDeclaration():
+        declarations.add(_parseFunctionDeclaration(member, lineInfo, path));
+        break;
+      case ast.TopLevelVariableDeclaration():
+        declarations.addAll(_parseVariableDeclaration(member, lineInfo, path));
+        break;
+      // Handle other types of declarations if needed
+      // Left for upcoming issues
+      // ex: MixinDeclaration, ExtensionDeclaration, EnumDeclaration, TypeAlias
+      default:
+        break;
+    }
+  }
+
+  return declarations;
+}
+
+// handle both TopLevelVariableDeclaration & FieldDeclaration
+List<Declaration> _parseVariableDeclaration(
+  dynamic declaration,
+  LineInfo lineInfo,
+  String path, {
+  Declaration? parent,
+}) {
+  // Variable Declaration might contain multiple variables
+  // (ex: int x, y, z = 1), then all the variables will have the same fields
+  // except for the variable id which is unique for each variable.
+
+  final sourceCode = declaration.toSource();
+  final startLine = lineInfo.getLocation(declaration.offset).lineNumber;
+  final endLine = lineInfo.getLocation(declaration.end).lineNumber;
+  final comments = _extractComment(declaration.documentationComment);
+  final declarations = <Declaration>[];
+
+  final ast.VariableDeclarationList variableList =
+      declaration is ast.TopLevelVariableDeclaration
+          ? declaration.variables
+          : declaration.fields;
+
+  for (final variable in variableList.variables) {
+    final id = variable.declaredFragment?.element.id ?? failedToExtractID;
+    declarations.add(
+      Declaration(
+        id,
+        name: variable.name.lexeme,
+        sourceCode: sourceCode,
+        startLine: startLine,
+        endLine: endLine,
+        path: path,
+        comment: comments,
+        parent: parent,
+      ),
+    );
+  }
+
+  return declarations;
+}
+
+// handle both FunctionDeclaration & MethodDeclaration
+Declaration _parseFunctionDeclaration(
+  dynamic declaration,
```

Comment:
You should pretty much never use `dynamic`. I think the only legitimate use for it is interacting with stuff like JSON.

The `ast.Declaration` class is a super type of both `ast.MethodDeclaration` and `ast.FunctionDeclaration` and has everything you need to construct your `Declaration` class except the name. So you could take the name as an extra string arg in `_parseFunctionDeclaration`.  Though if the name is mentioned in the `sourceCode` field, you might not even need it?

Reply:
In fact, I think a bunch of these parsers can be factorized. The only thing that really varies between the different declarations is what their name is, and what their parents/dependents are.